File: TypeScript/438.找到字符串中所有字母异位词.ts
function findAnagrams(s: string, p: string): number[] {
    const ans: Array<number> = new Array<number>()
    if (s.length < p.length) {
        return ans
    }
    const sts: Array<number> = new Array<number>(26).fill(0)
    const stp: Array<number> = new Array<number>(26).fill(0)
    for (let i = 0; i < p.length; ++i) {
        ++sts[s.charCodeAt(i) - 'a'.charCodeAt(0)]
        ++stp[p.charCodeAt(i) - 'a'.charCodeAt(0)]
    }
    if (JSON.stringify(sts) === JSON.stringify(stp)) {
        ans.push(0)
    }
    let left: number = 0;
    while (left + p.length < s.length) {
        --sts[s.charCodeAt(left) - 'a'.charCodeAt(0)]
        ++sts[s.charCodeAt(left + p.length) - 'a'.charCodeAt(0)]
        ++left
        if (JSON.stringify(sts) === JSON.stringify(stp)) {
            ans.push(left)
        }
    }
    return ans
};

File: Golang/438.找到字符串中所有字母异位词.go
package leetcode

func findAnagrams(s string, p string) []int {
	var ans []int
	if len(s) < len(p) {
		return ans
	}
	var sts [26]int
	var stp [26]int
	for i := 0; i < len(p); i++ {
		sts[s[i]-'a']++
		stp[p[i]-'a']++
	}
	if sts == stp {
		ans = append(ans, 0)
	}
	left := 0
	for left+len(p) < len(s) {
		sts[s[left]-'a']--
		sts[s[left+len(p)]-'a']++
		left++
		if sts == stp {
			ans = append(ans, left)
		}
	}
	return ans
}
